pkg/dyninst/rcjson/rcjson.go

```diff
@@ -213,9 +220,88 @@ type LogProbeCommon struct {
 	// Template is the message template of the log to emit.
 	Template string `json:"template"`
 	// Segments are the segments of the log message template.
-	Segments []json.RawMessage `json:"segments"`
+	Segments SegmentList `json:"segments"`
+}
+
+// TemplateSegment is a segment of a probe template.
+type TemplateSegment interface {
+	TemplateSegment() // marker method
+}
+
+// SegmentList is a list of Segments which can each be either a StringSegment or a JSONSegment
+type SegmentList []TemplateSegment
+
+// UnmarshalJSON implements custom JSON unmarshaling for SegmentList
+func (sl *SegmentList) UnmarshalJSON(data []byte) error {
+	var segmentData []json.RawMessage
+	if err := json.Unmarshal(data, &segmentData); err != nil {
+		return err
+	}
+
+	*sl = make([]TemplateSegment, len(segmentData))
+	for i, data := range segmentData {
+		var str StringSegment
+		if err := json.Unmarshal(data, &str); err == nil {
+			(*sl)[i] = str
+			continue
+		}
```

Comment:
what is this? this isn't how string segments are encoded, and you have support for `raw["str"]` below.